### ISSUE TITLE
System include extraction failed on Windows MSYS2 MinGW-w64 g++

### ISSUE THREAD

**Issue by Wybxc** [clangd/clangd#1106]
![image](https://user-images.githubusercontent.com/25005856/163089366-6f21c937-41da-49a4-9692-2498f2a0ace5.png)

**Logs**

```
I[10:39:57.280] <-- textDocument/documentLink(2)
E[10:39:57.451] System include extraction: driver execution failed with return code: 1 - ''. Args: ["D:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe" -E -x c++ - -v]
I[10:39:57.452] ASTWorker building file e:/xmake/hello-xmake/src/main.cpp version 1 with command 
[E:\xmake\hello-xmake]
"D:\\msys64\\mingw64\\bin/x86_64-w64-mingw32-g++.exe" --target=x86_64-w64-mingw32 --driver-mode=g++ -c -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG -o "build\\.objs\\hello-xmake\\windows\\x64\\release\\src\\main.cpp.obj" -resource-dir=D:/msys64/clang64/lib/clang/13.0.1 -- e:/xmake/hello-xmake/src/main.cpp
V[10:39:57.454] Driver produced command: cc1 -cc1 -triple x86_64-w64-windows-gnu -fsyntax-only -disable-free -disable-llvm-verifier -discard-value-names -main-file-name main.cpp -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fmath-errno -fno-rounding-math -mconstructor-aliases -mms-bitfields -munwind-tables -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb "-fcoverage-compilation-dir=E:\\xmake\\hello-xmake" -resource-dir D:/msys64/clang64/lib/clang/13.0.1 -D NDEBUG -internal-isystem "D:\\msys64\\mingw64/x86_64-w64-mingw32/include/c++/v1" -internal-isystem "D:\\msys64\\mingw64/include/c++/v1" -internal-isystem D:/msys64/clang64/lib/clang/13.0.1/include -internal-isystem "D:\\msys64\\mingw64/x86_64-w64-mingw32/include" -internal-isystem "D:\\msys64\\mingw64/include" -O3 -fdeprecated-macro "-fdebug-compilation-dir=E:\\xmake\\hello-xmake" -ferror-limit 19 -fvisibility hidden -fvisibility-inlines-hidden -fno-use-cxa-atexit -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -exception-model=seh -vectorize-loops -vectorize-slp -faddrsig -x c++ e:/xmake/hello-xmake/src/main.cpp
```

**`compile_commands.json`(generated by xmake)**
```
[
  {
    "directory": "E:\\xmake\\hello-xmake",
    "arguments": [
      "D:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe",
      "-c",
      "-fvisibility=hidden",
      "-fvisibility-inlines-hidden",
      "-O3",
      "-DNDEBUG",
      "-o",
      "build\\.objs\\hello-xmake\\windows\\x64\\release\\src\\main.cpp.obj",
      "src\\main.cpp"
    ],
    "file": "src\\main.cpp"
  }
]
```

**System information**

Output of `clangd --version`: 
```
clangd version 13.0.1
Features: windows
Platform: x86_64-w64-windows-gnu
```

Editor/LSP plugin: VSCode clangd v0.1.15

Operating system: Windows 11 21H2 22000

**Comment by Wybxc** [comment]
When executing clangd generated command line `D:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe -E -x c++ - -v` in shell, it got all infomation and returns 1:
```
Using built-in specs.
COLLECT_GCC=D:\msys64\mingw64\bin\g++.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-11.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/include --libexecdir=/mingw64/lib --enable-bootstrap --enable-checking=release --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-lto --enable-libgomp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev10, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' LDFLAGS_FOR_TARGET=-pipe --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Rev10, Built by MSYS2 project)
COLLECT_GCC_OPTIONS='-E' '-v' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
 D:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/cc1plus.exe -E -quiet -v -iprefix D:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.2.0/ -D_REENTRANT - -mtune=generic -march=x86-64 -dumpbase -
```

After removing the `-`, execute `D:\\msys64\\mingw64\\bin\\x86_64-w64-mingw32-g++.exe -E -x c++ -v`, it is correct and returns 0:
```
Using built-in specs.
COLLECT_GCC=D:\msys64\mingw64\bin\g++.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-11.2.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/include --libexecdir=/mingw64/lib --enable-bootstrap --enable-checking=release --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --disable-libstdcxx-debug --enable-lto --enable-libgomp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev10, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --with-boot-ldflags='-pipe -Wl,--disable-dynamicbase -static-libstdc++ -static-libgcc' LDFLAGS_FOR_TARGET=-pipe --enable-linker-plugin-flags='LDFLAGS=-static-libstdc++\ -static-libgcc\ -pipe\ -Wl,--stack,12582912'
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Rev10, Built by MSYS2 project)
```

**Comment by jives** [comment]
I have the same issue on Windows 10 with clangd 17.0.3 (using neovim).
If I specify clang as the compiler, everything works as expected. If I specify gcc, I see the same issue as in https://github.com/clangd/clangd/issues/1106#issuecomment-1097494830.

**Comment by HighCommander4** [comment]
> Maybe "-" should be replaced with "nul" on Windows, as indicated here https://clangd.llvm.org/guides/system-headers#heuristic-search-for-system-headers?

Do you understand how "-" causes this error?

The error clearly doesn't happen for all Windows users using `--query-driver`, since based on previous issues many Windows users are using `--query-driver` successfully (including with MinGW), so it would be good to understand what is different about the affected cases.

**Comment by hedger** [comment]
For anyone who stumbles upon this issue on Windows, besides checking that gcc binary itself is executable (by simply starting it from the explorer), also check that its backend - cc1.exe - is functional. It can be found at ..\libexec\gcc\arm-none-eabi\<version>\cc1.exe. 

In my case, I was getting the exact same error from clangd. Running the gcc binary with `-E -x c++ - -v` params only worked from `bin` folder of the toolchain, but not any other. The reason is that current work directory is also used on Windows for .dll lookup, and cc1.exe was depending on `libwinpthread-1.dll` which was only in `bin` folder. Copying `libwinpthread-1.dll` into the folder with cc1.exe resolved the issue for me.

To sum it up, it was a toolchain issue, not clangd's.

**Comment by Xeverous** [comment]
Looks like I have the same issue. I have `clangd version 19.1.2 (https://github.com/llvm/llvm-project 7ba7d8e2f7b6445b60679da826210cdde29eaf8b)` with MinGW on Windows 10 and the query driver just doesn't work:

```
E[23:52:07.082] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++-header -]
E[23:52:07.083] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++-header -]
E[23:52:07.083] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++ -]
E[23:52:07.083] System include extraction: driver execution failed with return code: 1 - ''. Args: ["C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++-header -]
```

The paths are correct and when I paste this command into cmd it does work and lists a lot of stuff. It works for both `g++` and `c++` But for some reason if clang does it, it returns 1.

The only interesting thing that I found that if I paste it into cmd, the results are printed but the shell prompt does not appear again. I need to press ctrl+C, which probably invalides application's return code by killing it.

If I write a command like `"C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++-header nul` it prints the same stuff but returns 0 and does not hang the shell prompt. @HighCommander4 can this be an issue? That clang does not terminate the command execution correctly?

**Comment by HighCommander4** [comment]
> The only interesting thing that I found that if I paste it into cmd, the results are printed but the shell prompt does not appear again. I need to press ctrl+C, which probably invalides application's return code by killing it.
> 
> If I write a command like `"C:\\mingw-w64-gcc-mcf-20250603-15.1.1-x64-msvcrt\\mingw64\\bin\\c++.exe" -E -v -x c++-header nul` it prints the same stuff but returns 0 and does not hang the shell prompt. [@HighCommander4](https://github.com/HighCommander4) can this be an issue? That clang does not terminate the command execution correctly?

I think these observations are expected.

The `-` argument tells `c++.exe` to read input from [stdin](https://en.wikipedia.org/wiki/Standard_streams#Standard_input_(stdin)). When you run that from the command line, `stdin` is hooked up to produce the characters you type on the command line, and the program reading from it does not get to the end until you enter an explicit "end of file" marker with <kbd>Ctrl</kbd>+<kbd>D</kbd> (or interrupt the program in some way, such as with <kbd>Ctrl</kbd>+<kbd>C</kbd>).

By contrast, `nul` is a special file that's empty, i.e. the first character the program tries to read from it is the end-of-file character.

The reason `-` should be fine for clangd's usage, is that it redirects the invoked child process's `stdin` to `nul`. (This happens by passing an empty string as the first element of `Redirects` [here](https://searchfox.org/llvm/rev/89f692a24f6a13ae5cf9e37f91abe6f34c403258/clang-tools-extra/clangd/SystemIncludeExtractor.cpp#338). Per the [documentation](https://searchfox.org/llvm/rev/89f692a24f6a13ae5cf9e37f91abe6f34c403258/llvm/include/llvm/Support/Program.h#116-123) of `ExecuteAndWait()`, "When an empty path is passed in, the corresponding file descriptor will be disconnected (ie, `/dev/null`'d) in a portable way." And the Windows implementation of that [handles that](https://searchfox.org/llvm/rev/89f692a24f6a13ae5cf9e37f91abe6f34c403258/llvm/lib/Support/Windows/Program.inc#142-143) by redirecting to `NUL`.)

As an experiment, I tried replacing the first element of `Redirects` there with `std::nullopt`, so that we get no redirection instead. When running `clangd` on the command line, e.g. with `clangd --query-drive=... --check=...`, the resulting behaviour is that `clangd` now blocks on me typing input until I press <kbd>Ctrl</kbd>+<kbd>D</kbd>.

Based on this, I suspect the issue is not likely to be related to the `-`. (If you're interested in building clangd locally with `-` changed to `nul` on an affected machine to confirm, that would give us more certainty on this point.)

I think the issue is more likely to be related to DLL lookup, as described in [this comment](https://github.com/clangd/clangd/issues/1106#issuecomment-2395519536) (but that's mostly a guess on my part).

**Comment by Xeverous** [comment]
I have tried manually querying the compiler from clangd's working directory (reported in logs) and it worked while clangd has still the same problem. Is there any way I can get better error than just negative return status?

**Comment by HighCommander4** [comment]
> Is there any way I can get better error than just negative return status?

I'm not sure.

Note that your logs show a return status of 1, which is positive rather than negative. According to [this comment](https://searchfox.org/llvm/rev/b73720cf6c5380854bf27d4453abf21cc87ae642/llvm/include/llvm/Support/Program.h#101-104):

>  /// \returns an integer result code indicating the status of the program.
>  /// A zero or positive value indicates the result code of the program.
>  /// -1 indicates failure to execute
>  /// -2 indicates a crash during execution or timeout

that suggests that the launching of the child process itself succeeded, but the child process then exited with a return value of 1 indicating failure.

**Comment by Xeverous** [comment]
> the child process then exited with a return value of 1 indicating failure

Then this makes it even more weird...

Can clangd print subprocess stdout and/or stderr? Is the return status the only thing available for diagnosing the problem?

**Comment by HighCommander4** [comment]
> Can clangd print subprocess stdout and/or stderr?

It could, by redirecting stdout and/or stderr to a file and then reading that file to see what, if anything, was written.

That's actually how clangd gets at the output to parse in the case of successful execution; the redirection is set up [here](https://searchfox.org/llvm/rev/77941eba7f01fc6576b3e060a3fb9cad1a64f9ea/clang-tools-extra/clangd/SystemIncludeExtractor.cpp#339) and the file is then read [here](https://searchfox.org/llvm/rev/77941eba7f01fc6576b3e060a3fb9cad1a64f9ea/clang-tools-extra/clangd/SystemIncludeExtractor.cpp#352). However, current the file is only read in case of successful execution.

If you're interested in doing some local experimentation, you could tweak the above places to read the file even in case of failed execution and see whether anything was written there (happy to provide more specific pointers if you'd like).

**Comment by Xeverous** [comment]
I can't build clangd at the moment but I could test such a build where it always prints what the compiler produced. I think it could be a worthwhile addition to always print compiler stdout/stderr into clangd's log.

**Comment by toneoesa** [comment]
This [comment](https://github.com/clangd/clangd/issues/1106#issuecomment-2395519536) is important. It solved the problem for me. Also, adding the bin directory of gcc.exe to PATH worked.

**Comment by HighCommander4** [comment]
Until a proper solution to this issue is found, a workaround is to perform manual configuration equivalent to what `--query-driver` does automatically. I wrote up some instructions for how to do that [here](https://gist.github.com/HighCommander4/8300f19f846cc99b462b11b35971bf97).

**Comment by fengxi-mao-mchp** [comment]
Hi @HighCommander4. On my own computer, clangd's current behavior of passing `-` as the input source file into the command to query g++ works fine. Here is the command it runs:
`g++.exe -E -x c++ - -v`

However, on two of my coworkers' computers, this exact same command results in the following output in the clangd log:

```
E[08:15:28.828] System include extraction: start marker not found: Using built-in specs.
COLLECT_GCC=C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/g++.exe
Target: x86_64-w64-mingw32
Configured with: ...
...
C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/../libexec/gcc/x86_64-w64-mingw32/8.5.0/cc1plus.exe -E -quiet -v -iprefix C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.5.0/ -D_REENTRANT - -mtune=core2 -march=nocona
```

Changing the stdin flag to `nul` in the clangd source code resolved this issue, and clangd was able to query g++ successfully.

I just wanted to ask: if `ExecuteAndWait` redirects stdin to `nul` anyway on Windows, why don’t we just pass in `nul` as the input source file instead of `-`? Was there a reason clangd chose to use `-`?

**Comment by HighCommander4** [comment]
> However, on two of my coworkers' computers, this exact same command results in the following output in the clangd log:
> 
> ```
> E[08:15:28.828] System include extraction: start marker not found: Using built-in specs.
> COLLECT_GCC=C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/g++.exe
> Target: x86_64-w64-mingw32
> Configured with: ...
> ...
> C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/../libexec/gcc/x86_64-w64-mingw32/8.5.0/cc1plus.exe -E -quiet -v -iprefix C:/Users/c78143/installers/features/PolarFireSystemDesigner/backend/hls/dependencies/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.5.0/ -D_REENTRANT - -mtune=core2 -march=nocona
> ```
> 
> Changing the stdin flag to `nul` in the clangd source code resolved this issue, and clangd was able to query g++ successfully.

Do you understand, or are you able to investigate, why using `nul` has this effect?

> I just wanted to ask: if `ExecuteAndWait` redirects stdin to `nul` anyway on Windows, why don’t we just pass in `nul` as the input source file instead of `-`? Was there a reason clangd chose to use `-`?

I think the original reason is that the spelling of the null device varies by platform (e.g. `nul` on Windows vs. `/dev/null` on Linux), and since we already had the platform-specific logic inside the implementation of `ExecuteAndWait`, the authors didn't want to duplicate it in clangd application code.

I'm not opposed to changing clangd to use `nul` directly on Windows, but I would feel better about it if we understood the explanation for the current behaviour.

**Comment by fengxi-mao-mchp** [comment]
I can’t seem to reproduce the issue anymore. It’s likely an environment issue, but I can’t pinpoint the exact cause.

**Comment by Xeverous** [comment]
Good news from me, I can no longer reproduce it aswell. Did not change clangd at all. All I did was to update VS Code. Now running on v1.103.2 and everything works. So it could be a VS Code issue after all.

**Comment by HighCommander4** [comment]
> In my case, I was getting the exact same error from clangd. Running the gcc binary with `-E -x c++ - -v` params only worked from `bin` folder of the toolchain, but not any other. The reason is that current work directory is also used on Windows for .dll lookup, and cc1.exe was depending on `libwinpthread-1.dll` which was only in `bin` folder. Copying `libwinpthread-1.dll` into the folder with cc1.exe resolved the issue for me.

See also https://github.com/clangd/vscode-clangd/issues/861 with more details about how this could be a DLL lookup issue.

If there really is a difference between how DLL lookup happens on Windows when a program is executed from the command-line, versus when clangd is executing it with `llvm::sys::ExecuteAndWait()`, then that's something we should consider fixing. (But it will take a Windows user who is willing to build clangd locally and experiment / prepare a patch to do so.)
